frontend/benefit/applicant/src/components/applications/forms/application/utils.ts

```diff
@@ -1,8 +1,42 @@
 import { Application } from 'benefit-shared/types/application';
 import { FormikErrors, FormikTouched } from 'formik';
 
-export const getTouchedAndInvalidFields = (
+type FlattenedObject = Record<string, string>;
+
+function flattenObject(
+  obj: FlattenedObject,
+  parentKey = '',
+  result: FlattenedObject = {}
+): FlattenedObject {
+  const appendResult = result;
+  Object.keys(obj).forEach((key) => {
+    const newKey = `${parentKey}${parentKey ? '.' : ''}${key}`;
+    if (
+      typeof obj[key] === 'object' &&
+      obj[key] !== null &&
+      !Array.isArray(obj[key])
+    ) {
+      flattenObject(obj[key] as unknown as FlattenedObject, newKey, result);
+    } else {
+      appendResult[newKey] = obj[key];
+    }
+  });
+  return result;
+}
+
+/**
+ * Used to enable / disable "save and close" button.
+ * Flatten Formik's error and touched fields to compare touched substring against error fields.
+ * Error fields are more verbose than touched fields, including min, max etc. validation data
+ * for the whole nested data set, for example, company or employee. Touch fields only consist
+ * of single fields that are touched. Thus substring is used to match touched fields against error fields.
+ * */
+export const findIntersectionOfTouchedAndErroredFields = (
   touched: FormikTouched<Partial<Application>>,
   errors: FormikErrors<Partial<Application>>
 ): string[] =>
-  Object.keys(touched).filter((field) => Object.keys(errors).includes(field));
+  Object.keys(flattenObject(touched as FlattenedObject)).filter((field) =>
+    Object.keys(flattenObject(errors as FlattenedObject)).some((errorField) =>
+      errorField.includes(field)
```

Comment:
With the above suggestion, these could just be passed as is.

Reply:
Oh, and `flattenObject` is being run for errors quite a few times if there are many touched fields. I doubt there's much of a reason to fear performance issues here, but the flattened objects could also be optionally assigned to variables first before this two-level lookup happens.